File: 14-1.rb
#!/usr/bin/env ruby

class Polymer
  attr_accessor :molecule
  attr_accessor :rules

  def initialize(template)
    @molecule = template
    @rules = Hash.new ''
  end

  def add_rule(pair, insert)
    @rules[pair] = insert
  end

  def step
    new_molecule = @molecule[0]
    (@molecule.size - 1).times do |pos|
      pair = @molecule[pos, 2]
      new_molecule += @rules[pair] + pair[1]
    end
    @molecule = new_molecule
  end

  def score
    counts = Hash.new 0
    @molecule.each_char do |atom|
      counts[atom] += 1
    end
    l = counts.values.sort
    l[-1] - l[0]
  end

  def to_s
    "<#{self.class}: #{@molecule}>"
  end

  def inspect
    to_s
  end
end

insertion_format = /^(?<pair>[[:upper:]]{2}) -> (?<insert>[[:upper:]])$/

input = File.read('14.input').lines.map(&:strip)

p = Polymer.new input.shift

input.shift

input.each do |line|
  insertion = insertion_format.match line
  p.add_rule insertion['pair'], insertion['insert']
end

10.times do
  p.step
end

print p.score, "\n"

File: 14-1.py
#!/usr/bin/env python

from collections import deque
import re

class Polymer:
    def __init__(self, template):
        self.molecule = template
        self.rules = {}

    def add_rule(self, pair, insert):
        self.rules[pair] = insert

    def step(self):
        new_molecule = self.molecule[0]
        for pos in range(len(self.molecule) - 1):
            pair = self.molecule[pos : pos + 2]
            new_molecule += self.rules[pair] + pair[1]
        self.molecule = new_molecule

    def score(self):
        counts = {}
        for atom in self.molecule:
            counts[atom] = counts.get(atom, 0) + 1
        l = list(counts.values())
        l.sort()
        return l[-1] - l[0]

    def __str__(self):
        return f"<{self.__class__.__name__}: {self.molecule}>"


insertion_format = re.compile("(?P<pair>[A-Z]{2}) -> (?P<insert>[A-Z])")

input = deque([line.strip() for line in open('14.input').readlines()])

p = Polymer(input.popleft())

input.popleft()

for line in input:
    insertion = insertion_format.match(line)
    p.add_rule(insertion['pair'], insertion['insert'])

for _ in range(10):
    p.step()

print(p.score())
